### DISCUSSION TITLE
Crazyflie Controller Question

### DISCUSSION THREAD

**Discussion by GNC-JY** [bitcraze/discussions#922] [Q&A] [answered]
Hello.

I'm using the crazyflie 2.1.

I had a question while analyzing crazyflie firmware.

While analyzing the attitude controller, a question arose with rate control.

rate desired uses the deg value calculated by the attack control.
As a result of log data that the value from sensors->gyro used as rate actual, it seemed that the value with a duration of -35 to 35 deg/s. In addition, the rate desired value also seemed to have a saturation of -1080 to 1080.

I want to know the code part that has saturation on this part. And I want to know why it's like that. And rollout, pitchout, and yawout that have passed through rate control are used as power distribution, and I want to know the reason for the 16 bit saturation.

best regards.

**Comment by krichardsson** [comment]
Hi, 
I'm not completely sure which part of the code you are referring to, could you please attach links to the code to clarify. 

The easiest way to do this is to open the [file in github](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/controller_pid.c). Click the line number of the start of the code block you are interested in (it will turn yellow), shift click the line number of the end of the code block to mark it. Click the "..." icon and chose "Copy permalink", you will ge something like this https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/src/modules/src/controller/controller_pid.c#L18-L25

**Comment by GNC-JY** [comment]
https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/src/modules/src/controller/controller_pid.c#L124-L125

Here is the code parts.

**Comment by knmcguire** [comment]
Hi! 

So for the [saturating int16](https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/src/modules/src/controller/attitude_pid_controller.c#L146 
) part, the reasoning is simple. You don't want to that value to overflow over the int16 maximum value or else it will interpreted from the minimum value (so it flips). That will cause some really bad behavior and you don't want that for sure. Why we didn't choose an int32 to counteract that, I also would find it very unlikely to see an attitude rate go over 32768 deg/sec, but in embedded programming is always good to be safe than sorry I guess? Also I think at that time it was just choice by design but we would have implemented this now, I think we probably would have added an assert instead as those values are unrealistic and there is obvious something is wrong if it overflows.

About the other capping of the values, I can not find in the PID code that the gyro values or the desired rate values are forcefully saturated specifically. Only that the final PID output is limited. My guess is that it happens earlier in the [stabilizer loop phase](https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/src/modules/src/stabilizer.c#L319), so when the setpoint and the sensors are acquired. This requires more digging into the code that will take a bit more time so I'll leave that to you, so let us know if you have any questions about that. Just note that limiting these types of values is not an uncommon practise with PID controllers (however it is a bit weird that the desired values and sensor values would be limited maybe.

**Comment by GNC-JY** [comment]
Hello.
I'm currently experimenting with rapid maneuvers with crazyflie (e.g. quadcopter throwing). The saturation part is the part that raised the question during the simulation and actual experiment.
Rate desired is measured only 1080 deg/s in extreme rotations, and values from sensors->gyro are measured only 35 deg/s in extreme rotations. (I used a complementary filter).
![image](https://github.com/bitcraze/discussions/assets/110527007/e4602a36-ac4a-4b81-b2be-a36312a30b84)
![image](https://github.com/bitcraze/discussions/assets/110527007/c88b4bfb-ce41-42e3-bff7-3199dc818fde)
![image](https://github.com/bitcraze/discussions/assets/110527007/823684d2-f0cf-4a92-b399-385f91f6d7a1)

r_pitch  parameter is sensors->gyro.y.
pitchRate parameter is rateDesired.pitch.

Even if desired is limited, it is questionable why the value of the gyro sensor takes saturation. I think it can't be helped if it's the limit of the sensor.

**Reply by knmcguire** [reply, marked as answer]
The desiredRate is the setpoint that comes directly from the rate PID controller (if you don't use rate control from the controller), so it might also be a factor of either the intergral limit or something else. As this is depended to many factors it is difficult to see where this comes from by just looking it from the eye and since this is not a blocking bug. If you'd really like to go into it, there are p[ython bindings](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/building-and-flashing/build/#build-python-bindings) available of the controllers, where you give give some inputs to it and see what the result is. That might be easier to figure out where this limit comes from. 

About the limit gyro, it might even be configured as low level as the registries of the settings of the sensor... this will take a bit too much time for us to dig in for you but I can point you in the right direction:

This is the driver for the gyro: https://github.com/bitcraze/crazyflie-firmware/blob/master/src/drivers/bosch/src/bmi055_gyro.c

**Reply by GNC-JY** [reply]
I am also currently analyzing the code as to why the value coming from the gyro sensor has a saturation of 35deg/s.
There is no speculation in gyro driver.
Which part of the code stores sensor values in sensors->gyro?

**Reply by knmcguire** [reply]
https://github.com/bitcraze/crazyflie-firmware/blob/cd3f3dfb668261f4461f972be8116d996aa7256e/src/hal/src/sensors_bmi088_bmp388.c#L273

**Reply by GNC-JY** [reply]
I think the problem with this question has been solved. I found the answer in stabilizer.c. (sensorData.gyro)

**Reply by knmcguire** [reply]
Great to hear. Could you specify the location?
